src/solver/lu_decomposition.cu

```diff
@@ -6,193 +6,174 @@
+            cudaMemcpy(d_uik_nkj, solver.uik_nkj_, sizeof(std::pair<size_t, size_t>)*solver.uik_nkj_size_, cudaMemcpyHostToDevice);
+            cudaMemcpy(d_lij_ujk, solver.lij_ujk_, sizeof(std::pair<size_t, size_t>)*solver.lij_ujk_size_, cudaMemcpyHostToDevice);
+            cudaMemcpy(d_lki_nkj, solver.lki_nkj_, sizeof(std::pair<size_t, size_t>)*solver.lki_nkj_size_, cudaMemcpyHostToDevice);
+            cudaMemcpy(d_lkj_uji, solver.lkj_uji_, sizeof(std::pair<size_t, size_t>)*solver.lkj_uji_size_, cudaMemcpyHostToDevice);
+            cudaMemcpy(&(device->A_),&d_A, sizeof(double*), cudaMemcpyHostToDevice);
+            cudaMemcpy(&(device->L_),&d_L, sizeof(double*), cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->U_),&d_U, sizeof(double*), cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->do_aik_), &d_do_aik, sizeof(char*), cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->aik_), &d_aik, sizeof(size_t*), cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->do_aki_),&d_do_aki,sizeof(char*),cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->aki_),&d_aki, sizeof(size_t*), cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->uii_), &d_uii, sizeof(size_t*), cudaMemcpyHostToDevice);
+            cudaMemcpy(&(device->niLU_), &d_niLU, sizeof(std::pair<size_t, size_t>*), cudaMemcpyHostToDevice);
+            cudaMemcpy(&(device->uik_nkj_), &d_uik_nkj, sizeof(std::pair<size_t, size_t>*), cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->lij_ujk_), &d_lij_ujk, sizeof(std::pair<size_t, size_t>*), cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->lki_nkj_), &d_lki_nkj, sizeof(std::pair<size_t, size_t>*), cudaMemcpyHostToDevice); 
+            cudaMemcpy(&(device->lkj_uji_), &d_lkj_uji, sizeof(std::pair<size_t, size_t>*), cudaMemcpyHostToDevice); 
+        
+            //total number of threads is number of blocks in sparseMatrix A 
+            size_t num_block = (sparseMatrix.n_grids_ + BLOCK_SIZE - 1) / BLOCK_SIZE; 
+          
+           // call kernel
+            auto startTime = std::chrono::high_resolution_clock::now();
+            DecomposeKernel<<<num_block, BLOCK_SIZE>>>(device, sparseMatrix.n_grids_, solver.niLU_size_); 
+            cudaDeviceSynchronize();
+            auto endTime = std::chrono::high_resolution_clock::now();
+            auto kernel_duration = std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime);
+            cudaMemcpy(sparseMatrix.L_, d_L, sizeof(double)* sparseMatrix.L_size_, cudaMemcpyDeviceToHost); 
+            cudaMemcpy(sparseMatrix.U_, d_U, sizeof(double)* sparseMatrix.U_size_, cudaMemcpyDeviceToHost); 
+        //clean up 
+        // cudaFree(d_A); 
+        // cudaFree(d_L); 
+        // cudaFree(d_U); 
+        // cudaFree(d_do_aik); 
+        // cudaFree(d_aik);
+        // cudaFree(d_do_aki); 
+        // cudaFree(d_aki); 
+        // cudaFree(d_uii); 
+        // cudaFree(device); 
```

Comment:
Why you do not need to deallocate these memories any more?

Reply:
LinearSolve constructor called decompose function before calling solve() function in CudaLinearSolver(). I thought those data was needed for solve(). The test still passed after I deallocate these memories. I revised it now and deallocate them.

Reply:
Thanks.